### ISSUE TITLE
hca

### ISSUE THREAD

**Issue by reed-mcdaniel-716** [Bioconductor/Contributions#1926]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/Bioconductor/hca

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.
  
- [x] I am familiar with the [Bioconductor code of conduct][7] and 
  agree to abide by it.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/
[7]: https://bioconductor.org/about/code-of-conduct/

**Comment by vjcitn** [comment]
I think this package is in very good shape.  BiocCheck had no striking findings except perhaps
an instance of <<-, which should be avoided if possible.

The overall test coverage is good, but there is a lack of coverage of some
functionalities in the vignette.
```
> covr::package_coverage(type="vignettes")
hca Coverage: 70.02%
R/bundles.R: 0.00%
R/details.R: 0.00%
R/samples.R: 0.00%
R/summary.R: 0.00%
R/utilities.R: 60.00%
R/filters.R: 62.75%
R/files.R: 78.85%
R/tbl_hca.R: 82.86%
R/projects.R: 91.67%
R/hca.R: 94.44%
R/term_facets.R: 96.43%
R/index.R: 96.77%
R/lol.R: 99.10%
R/lol_hits.R: 100.00%
```
Perhaps this is fine, but if there is some functionality in the low-coverage files that
deserves to be discussed in vignette, consider adding some demonstrations.

Here's the overall:
```
> library(covr)
1/0 packages newly attached/loaded, see sessionInfo() for details.
> ?package_coverage
> covr::package_coverage(type="all")
hca Coverage: 96.91%
R/summary.R: 85.42%
R/utilities.R: 90.00%
R/hca.R: 94.44%
R/bundles.R: 95.83%
R/samples.R: 95.83%
R/term_facets.R: 96.43%
R/tbl_hca.R: 97.14%
R/files.R: 98.08%
R/lol.R: 99.10%
R/details.R: 100.00%
R/filters.R: 100.00%
R/index.R: 100.00%
R/lol_hits.R: 100.00%
R/projects.R: 100.00%
```

I ran pkgnet::CreatePackageReport("hca") to look at package and function dependencies.  The function
dependency layout had some islands that could be of interest

![Screenshot from 2021-03-03 12-07-53](https://user-images.githubusercontent.com/4624066/109843660-5068bc80-7c19-11eb-9763-bfe37cb13ab4.png)

My last comment concerns scope.  I suspect that the lol capabilities will be of interest outside the HCA
project.  They have to live somewhere, so why not there?  I found them very useful when probing json
outputs of diverse APIs.  And I think an R Journal article may be worth producing on this capability.

However, when I look a little deeper:
```
.lol_visit_impl.list <- function(x, path, index, dict) {
    dict[[path]] <- append(dict[[path]], list(index))
    attr(dict[[path]], "leaf") <- FALSE

    nms <- names(x)
    is_null_nms <- is.null(nms)
    if (is_null_nms) {
        ## nms <- paste0("[[", seq_along(x), "]]")
        nms <- rep("[*]", length(x))
    }
    if (identical(path, ".")) {
        path <- nms
    } else {
        if (is_null_nms) {
            path <- paste0(path, nms)
        } else {
            path <- paste0(path, ".", nms)
        }
    }

    for (i in seq_along(x))
        .lol_visit_impl(x[[i]], path[[i]], append(index, i), dict)
}
```
with no commenting ... I wonder how maintainable this will be?

In summary I think the package is close to acceptance, but some consideration
should be given to these points.

**Comment by vjcitn** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/reed-mcdaniel-716.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("hca")`. The package 'landing page' will be created at

https://bioconductor.org/packages/hca

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/
